### PR TITLE
Add fuzzer based on honggfuzz

### PR DESCRIPTION
**Pull request by PsiACE**
fuzzing

Also see: #211

### CODE REVIEW

**Comment by BohuTANG** [comment]
Hi, @Dandandan @andygrove 

With this patch, we can easily to run the fuzz tool for sqlparser-rs and find more fuzz issues, great for robustness.
In the future we can make a CI to run it (limited time, i.e 10 minutes).

File: src/parser.rs [hunk 1]
```diff
@@ -2358,8 +2358,7 @@ impl<'a> Parser<'a> {
                         ]) // This couldn't possibly be a bad idea
                     })?
                     .into_iter()
-                    .filter(|i| i.is_some())
-                    .map(|i| i.unwrap())
+                    .flatten()
```
Comment (alamb) [inline src/parser.rs:2361]: Is it correct that  this is a real bug that was found/fixed with the fuzzer?

File: fuzz/fuzz_targets/fuzz_parse_sql.rs [hunk 2]
```diff
@@ -0,0 +1,12 @@
+use honggfuzz::fuzz;
+use sqlparser::dialect::GenericDialect;
+use sqlparser::parser::Parser;
+
+fn main() {
+    loop {
+        fuzz!(|data: String| {
```
Comment (alamb) [inline fuzz/fuzz_targets/fuzz_parse_sql.rs:7]: This is pretty cool. Where does the data come from? Is it just random strings? Or does this somehow know how to make valid sql statements?

**Review by alamb — COMMENTED** [review]
Sorry for the delay in review. I am trying to pitch in and help out now to clear the backlog. 

This is very cool @PsiACE  and @BohuTANG  -- thank you. It is cool that it seems you have found a bug with this fuzzer already.

My biggest question is about the input -- does the fuzzer simply check against random inputs or does it somehow know how to randomly make sql statements?

File: fuzz/fuzz_targets/fuzz_parse_sql.rs (hunk 2 above)
Comment (alamb) [reply fuzz/fuzz_targets/fuzz_parse_sql.rs:7]: I see some related comments here: https://github.com/sqlparser-rs/sqlparser-rs/pull/211#issuecomment-650716166

**Review by alamb — APPROVED** [review]
I think it is cool and having initial fuzzing, even if just random strings to start, is a good first step.

File: src/parser.rs (hunk 1 above)
Comment (PsiACE) [reply src/parser.rs:2361]: no, just make clippy happy

File: fuzz/fuzz_targets/fuzz_parse_sql.rs (hunk 2 above)
Comment (PsiACE) [reply fuzz/fuzz_targets/fuzz_parse_sql.rs:7]: For now it is some random bytes, but can be used to find bugs. Optimization can be continued later, refer to: https://www.cockroachlabs.com/blog/sqlsmith-randomized-sql-testing/

Comment (alamb) [reply fuzz/fuzz_targets/fuzz_parse_sql.rs:7]: In a past project, we had a harness that could generate random SQL queries and it found **many** bugs -- such tests are a wonderful way to help database software mature

Comment (alamb) [reply fuzz/fuzz_targets/fuzz_parse_sql.rs:7]: Filed https://github.com/apache/arrow-datafusion/issues/913 for DataFusion. Thanks for the pointer @PsiACE

**Review by houqp — APPROVED** [review]
This is pretty cool. Looking forward to the day when we could have sql syntax specific fuzzer :)

**Pull request merged**